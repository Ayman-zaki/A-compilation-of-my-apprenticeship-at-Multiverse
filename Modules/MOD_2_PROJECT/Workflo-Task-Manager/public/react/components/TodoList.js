import React, { useEffect, useState } from "react";
import CreateTaskPopup from "./CreateTask";

const TodoList = () => {
  const [modal, setModal] = useState(false);

  const toggle = () => {
    setModal(!modal);
  };

  const [taskList, setTaskList] = useState([]);

  const saveTask = (taskObj) => {
    let tempList = taskList;
    tempList.push(taskObj);
    localStorage.setItem("taskList", JSON.stringify(tempList));
    setTaskList(taskList);
    setModal(false);
  };

  const [todos, setTodos] = useState([]);
  const [inprogress, setInprogress] = useState([]);
  const [completed, setCompleted] = useState([]);

  const [taskName, setTaskName] = useState("");
  const [description, setDescription] = useState("");
  const [assignedTo, setAssignedTo] = useState("");
  const [dueDate, setDueDate] = useState("");
  const [priorityLevel, setPriorityLevel] = useState("");

  const addTodo = () => {
    const todo = {
      id: Math.floor(Math.random() * 1000),
      taskName: taskName,
      description: description,
      assignedTo: assignedTo,
      dueDate: dueDate,
      priorityLevel: priorityLevel,
    };
    setTodos([todo, ...todos]);
    console.log(todo);
  };

  const addToProgress = (id) => {
    const item = todos.find((x) => x.id === id);
    setInprogress([item, ...inprogress]);

    const filterarray = todos.filter((x) => x.id !== id);
    setTodos(filterarray);
  };
  const deleteTodo = (id) => {
    const filterarray = todos.filter((x) => x.id !== id);
    setTodos(filterarray);
  };
  const deleteInProgress = (id) => {
    const filterarray = inprogress.filter((x) => x.id !== id);
    setInprogress(filterarray);
  };

  const addtoCompleted = (id) => {
    const item = inprogress.find((x) => x.id === id);
    setCompleted([item, ...completed]);

    const filterarray = inprogress.filter((x) => x.id !== id);
    setInprogress(filterarray);
  };
  useEffect(() => {}, [todos, inprogress]);

  const deleteCompleted = (id) => {
    const filterarray = completed.filter((x) => x.id !== id);
    setCompleted(filterarray);
  };

  return (
   <>
      <div className="header">
        <div className="row">
          <div className="col">
            <div className="taskbtn">
              <button
                className="btn btn-outline-dark btn-block rounded-pill" id="creatTaskBtn"
                onClick={() => setModal(true)}
              >
                Create a Task
              </button>
            </div>
          </div>
        </div>
      </div>

      <div className="container-fluid">
        <div className="todos_wrapper row" id="todoWrapper">
          <div className="todos_list col-sm-4" id="todo-card">
            <h3 className="todo_title todo">To Do</h3>
            {todos.map((item, index) => (
              <div className="todo_card" key={item.id}>
                <p className="card_text">{item.taskName}</p>
                <p className="card_text">{item.description}</p>
                <p className="card_text">{item.assignedTo}</p>
                <p className="card_text">{item.dueDate}</p>
                <p className="card_text">{item.priorityLevel}</p>
                <i className='bx bxs-check-square' onClick={() => addToProgress(item.id)}></i>
                <i className='bx bxs-trash' onClick={() => deleteTodo(item.id)}></i>
              </div>
            ))}
          </div>

          <div className="todos_list col-sm-4" id="inProgress-card">
            <h3 className="todo_title inprogress">In Progress</h3>
            {inprogress.map((item, index) => (
              <div className="progress_card" key={item.key}>
                <p className="card_text">{item.taskName}</p>
                <p className="card_text">{item.description}</p>
                <p className="card_text">{item.assignedTo}</p>
                <p className="card_text">{item.dueDate}</p>
                <p className="card_text">{item.priorityLevel}</p>
                <i className='bx bxs-check-square' onClick={() => addtoCompleted(item.id)}></i>
                <i className='bx bxs-trash' onClick={() => deleteInProgress(item.id)}></i>
              </div>
            ))}
          </div>
          <div className="todos_list col-sm-4" id="complete-card">
            <h3 className="todo_title complete">Completed</h3>
            {completed.map((item, index) => (
              <div className="completed_card" key={item.id}>
                <p className="card_text">{item.taskName}</p>
                <p className="card_text">{item.description}</p>
                <p className="card_text">{item.assignedTo}</p>
                <p className="card_text">{item.dueDate}</p>
                <p className="card_text">{item.priorityLevel}</p>
                <i className='bx bxs-trash' onClick={() => deleteCompleted(item.id)}></i>
              </div>
            ))}
          </div>
        </div>
      </div>
      <CreateTaskPopup
        toggle={toggle}
        modal={modal}
        save={saveTask}
        addTodo={addTodo}
        taskName={taskName}
        description={description}
        assignedTo={assignedTo}
        dueDate={dueDate}
        priorityLevel={priorityLevel}
        setTaskName={setTaskName}
        setDescription={setDescription}
        setAssignedTo={setAssignedTo}
        setDueDate={setDueDate}
        setPriorityLevel={setPriorityLevel}
      />
    </>
  );
};

export default TodoList;
