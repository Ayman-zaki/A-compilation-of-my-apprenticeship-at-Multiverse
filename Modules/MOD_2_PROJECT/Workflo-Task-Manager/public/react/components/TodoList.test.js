import React from "react";
import renderer from "react-test-renderer";
import TodoList from "./TodoList";
import CreateTaskPopup from "./CreateTask";

describe("Create Task component", () => {
  //Snapshot test: Does our Component render as expected?

  test("renders Todo List", () => {
    const component = renderer.create(<TodoList />);

    const snapshot = component.toJSON();

    console.log("What does our snapshot look like?  ", snapshot);

    expect(snapshot).toMatchSnapshot();
  });

    //Behavior Test: Does our Component handle data passed in as props?

  test("Contains items passed down as props", () => {
    const taskName = ['Filter', 'Search'];
    
    const component = renderer.create(<CreateTaskPopup  taskName={taskName}/>);
    const treeEl = component.toTree()

    

    expect(treeEl.props.taskName.length).toBe(2)
  })
});
