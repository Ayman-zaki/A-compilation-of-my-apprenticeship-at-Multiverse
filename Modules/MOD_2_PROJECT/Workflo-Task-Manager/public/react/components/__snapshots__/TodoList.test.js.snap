// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Task component renders Todo List 1`] = `
Array [
  <div
    className="header"
  >
    <h2
      className="title"
    >
      WORK
      <span
        className="flo"
      >
        FLO
      </span>
    </h2>
    <div
      className="row"
    >
      <div
        className="col"
      >
        <div
          className="taskbtn"
        >
          <button
            className="btn btn-outline-dark btn-block rounded-pill"
            id="creatTaskBtn"
            onClick={[Function]}
          >
            Create a Task
          </button>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="input-group"
          id="search-container"
        >
          <input
            aria-describedby="search-addon"
            aria-label="Search"
            className="form-control rounded"
            placeholder="Search for a task"
            type="search"
          />
          <button
            className="btn btn-outline-dark"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
      <div
        className="col"
      >
        <div
          className="select-header"
        >
          <select
            aria-label="select something"
            className="form-select"
          >
            <option
              defaultValue={true}
            >
              Filter By
            </option>
            <option
              value="1"
            >
              Priority
            </option>
            <option
              value="2"
            >
              Asignee
            </option>
            <option
              value="3"
            >
              Alphabetical
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="task-container"
  >
    <div
      className="todos_wrapper"
    >
      <div
        className="todos_list"
      >
        <h3
          className="todo_title todo"
        >
          To Do
        </h3>
      </div>
      <div
        className="todos_list"
      >
        <h3
          className="todo_title inprogress"
        >
          In Progress
        </h3>
      </div>
      <div
        className="todos_list"
      >
        <h3
          className="todo_title complete"
        >
          Completed
        </h3>
      </div>
    </div>
  </div>,
]
`;
